raiden/transfer/mediated_transfer/mediator.py

```diff
@@ -0,0 +1,890 @@
+            it's assumed that the available_routes list is ordered from best to
+            worst.
+        timeout_blocks (int): Base number of available blocks used to compute
+            the lock timeout.
+        block_number (int): The current block number.
+    """
+    assert timeout_blocks > 0
+    assert timeout_blocks <= payer_transfer.expiration - block_number
+
+    transfer_pair = None
+    mediated_events = list()
+
+    payee_route = next_route(
+        routes_state,
+        timeout_blocks,
+        payer_transfer.amount,
+    )
+
+    if payee_route:
+        assert payee_route.reveal_timeout < timeout_blocks
+
+        lock_timeout = timeout_blocks - payee_route.reveal_timeout
+        lock_expiration = lock_timeout + block_number
+
+        payee_transfer = LockedTransferState(
+            payer_transfer.identifier,
+            payer_transfer.amount,
+            payer_transfer.token,
+            payer_transfer.target,
+            lock_expiration,
+            payer_transfer.hashlock,
+            payer_transfer.secret,
+        )
+
+        transfer_pair = MediationPairState(
+            payer_route,
+            payer_transfer,
+            payee_route,
+            payee_transfer,
+        )
+
+        mediated_events = [
+            mediatedtransfer(payee_transfer, payer_route.node_address),
+        ]
+
+    return (
+        transfer_pair,
+        mediated_events,
+    )
+
+
+def set_secret(state, secret):
+    """ Set the secret to all mediated transfers.
+
+    It doesn't matter if the secret was learned through the blockchain or a
+    secret reveal message.
+    """
+    state.secret = secret
+
+    for pair in state.transfers_pair:
+        pair.payer_transfer.secret = secret
+        pair.payee_transfer.secret = secret
+
+
+def set_payee_state_and_check_reveal_order(transfers_pair,  # pylint: disable=invalid-name
+                                           payee_address,
+                                           new_payee_state):
+    """ Set the state of a transfer *sent* to a payee and check the secret is
+    being revealed backwards.
+
+    Note:
+        the elements from transfers_pair are changed in place, the list must
+        contain all the know transfers to properly check reveal order.
+    """
+    assert new_payee_state in MediationPairState.valid_payee_states
+
+    wrong_reveal_order = False
+    for back in reversed(transfers_pair):
+        if back.payee_route.node_address == payee_address:
+            back.payee_state = new_payee_state
+            break
+
+        elif back.payee_state not in STATE_SECRET_KNOWN:
+            wrong_reveal_order = True
+
+    if wrong_reveal_order:
+        # TODO: Append an event for byzantine behavior.
+        # XXX: With the current events_for_withdraw implementation this may
+        # happen, should the notification about byzantine behavior removed or
+        # fix the events_for_withdraw function fixed?
```

Comment:
I use the tag `TODO` as for something that is known that needs to be done, in the aforementioned tag I do not know what and if something should be done, either someone that is reading through the code should think about it and make a sound decision or it needs further discussion.

In the given case the list is empty because there is not Event for the byzantine behavior yet, it may be added and if it's then the `TODO` just above should be done.

I'm not sure if I understand what is a "critical log event", if it's an Event object, that would be the byzantine behavior event, if it's a python logging call then I don't think that is a good solution, I have purposely avoided IO in the whole of the PR.